sklearn/utils/validation.py

```diff
@@ -266,11 +270,19 @@ def check_array(array, accept_sparse=None, dtype=None, order=None, copy=False,
         accept_sparse = [accept_sparse]
 
     if sp.issparse(array):
+        if dtype == "numeric":
+            dtype = None
         array = _ensure_sparse_format(array, accept_sparse, dtype, order,
                                       copy, force_all_finite)
     else:
         if ensure_2d:
             array = np.atleast_2d(array)
+        if dtype == "numeric":
+            if getattr(array, "dtype", None) is np.dtype(object):
```

Comment:
Same thing here: I think that checking dtype.kind is more robust, in particular to future evolutions of numpy.